e2e_testing/torchscript/index_put.py

```diff
@@ -0,0 +1,56 @@
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+# Also available under a BSD-style license. See LICENSE.
+
+import torch
+
+from torch_mlir_e2e_test.torchscript.framework import TestUtils
+from torch_mlir_e2e_test.torchscript.registry import register_test_case
+from torch_mlir_e2e_test.torchscript.annotations import annotate_args, export
+
+# ==============================================================================
+
+
+class IndexPutImplOneDimensionalFloatModule(torch.nn.Module):
+
+  def __init__(self):
+    super().__init__()
+
+  @export
+  @annotate_args([
+      None,
+      ([10], torch.float32, True),
+      ([6], torch.int64, True),
+      ([6], torch.float32, True),
```

Comment:
Not sure why the tests are failing, but I did notice some strange behavior. If you set the sizes for the `index` and `value` in the `annotate_args` to zero, the tests pass! I also noticed the following:

Assume that `input = torch.zeros(10, dtype=torch.float)`, `index = torch.tensor([0,0,0], dtype=torch.int64)`, and `value = torch.tensor([3,4,5], dtype=torch.float)`. 

If you set the sizes for the `index` and `value` in `annotate_args` to:

- `0` then `output[0] = 3 + 4 + 5`
- `1` then `output[0] = (3 + 3) + 4 + 5`
- `2` then `output[0] = (3 + 3) + (4 + 4) + 5`
- `3` then `output[0] = (3 + 3) + (4 + 4) + (5 + 5)`

This is why your test is failing. For some reason every value gets doubled before getting added to the tensor. I can't really think of what could be causing this behavior. @cathyzhyi, have you seen anything like this?

Reply:
I am still investigating this.  Your experiment result looks interesting! It almost look like the testing run is reusing the tensor from the baseline run.

Reply:
yea, if you print the `input` here https://github.com/llvm/torch-mlir/blob/1a2a9e066f320e725927a7094b5a668eae358453/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/torchscript/configs/linalg_on_tensors_backend.py#L53 It shows the `input` of testing run contains the result from the baseline run already because the torchscript run would update the `input` in place.

Reply:
Cool! That also explains why setting the `yieldValue` to the original value was working in the case of no accumulation.

Reply:
What could be the solution to this? @ramiro050 @cathyzhyi

Reply:
As of now, can we go ahead with this temporary fix?

Reply:
Yeah, I think they would be fine

Reply:
We will eventually need to extend the testing framework to correctly check that inputs are updated in place. However, in Torch-MLIR we don't really have runtime support for that (note that we set `True` on the annotations saying that the input should be treated with value semantics, i.e. in-place updates are not externally visible from the compiled code). So I think the right solution for now is for the e2e test framework to clone inputs before passing them to torchscript or native_torch functions. I.e., I think the right fix at the moment is to clone tensor inputs in these two lines:

https://github.com/llvm/torch-mlir/blob/1a2a9e066f320e725927a7094b5a668eae358453/python/torch_mlir_e2e_test/torchscript/configs/torchscript.py#L31

https://github.com/llvm/torch-mlir/blob/1a2a9e066f320e725927a7094b5a668eae358453/python/torch_mlir_e2e_test/torchscript/configs/native_torch.py#L27

We could make Trace.inputs a private member Trace._inputs and then have a method `Trace.get_inputs()` which clones appropriately. Then we can remove this manual input.clone()

Can you do that in a follow-up patch?

Reply:
Okay, I will do this in a follow-up patch.